# **Walmart Coffee Listings from 500 stores**

## **Final Case Study**

By: Romero Isa

Source: https://www.kaggle.com/datasets/dimitryzub/walmart-coffee-listings-from-500-stores

The reason why I pick this dataset is because I love coffee and out of the datasets about coffee on Kaggle that I went through, this is the most well put together one.

### **Questions to Consider:**
1. What are the top 10 coffee sellers?
2. Which coffee is the most popular by number of reviews?
3. What's the weight distribution of the coffee bags sold in Walmart?
4. What's the price distribution the coffee sold in Walmart?
5. What's the breakdown of the top 20 coffee type that is being sold in Walmart?
6. What's the correlation between price and rating? (min. 100 reviews)
7. What are the most affordable coffee you can get in Walmart? (gramms per dollar value)
8. What are the average prices of the top 5 most popular coffee type?

### **Exploratory Data Analysis**

In [ ]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

In [ ]:
df_raw = pd.read_csv("https://github.com/romerorisa/Walmart-Coffee-Listings/raw/main/coffee-listings-from-all-walmart-stores.csv")
df_raw

title  \
0       folgers classic roast ground coffee, 40.3-ounce   
1     café bustelo, espresso style dark roast ground...   
2     folgers classic roast ground coffee, medium ro...   
3     maxwell house original roast ground coffee, 42...   
4     great value classic roast medium ground coffee...   
...                                                 ...   
1395  xhao reusable capsules easy to clean easy to u...   
1396  cafe yaucono espresso dark roast ground coffee...   
1397  starbucks dark roast coffee with 2x caffeine, ...   
1398  starbucks holiday blend, medium roast k-cup co...   
1399  super coffee vanilla latte iced coffee bottle,...   

                     coffee_type  rating  reviews  \
0                  classic roast     3.8       93   
1            espresso,dark roast     4.7      914   
2     medium roast,classic roast     4.4      740   
3                            NaN     4.8     1321   
4                  classic roast     4.7     1598   
...                          ...     ...      ...   
1395                         NaN     0.0        0   
1396         espresso,dark roast     0.0        0   
1397          arabica,dark roast     4.3     5036   
1398        arabica,medium roast     4.5      397   
1399           iced coffee,latte     4.3     2791   

                       seller_name  \
0                      walmart.com   
1                      walmart.com   
2                      walmart.com   
3                      walmart.com   
4                      walmart.com   
...                            ...   
1395              joybuy selection   
1396  coffee & filters direct, inc   
1397                   walmart.com   
1398                   walmart.com   
1399                   walmart.com   

                                              thumbnail  price         weight  \
0     https://i5.walmartimages.com/asr/1fbbd523-8554...  13.92     40.3-ounce   
1     https://i5.walmartimages.com/asr/99a53df0-0471...   3.76          10 oz   
2     https://i5.walmartimages.com/asr/e6aba325-608e...   9.97     25.9 ounce   
3     https://i5.walmartimages.com/asr/a5be9586-b75d...   9.92        42.5 oz   
4     https://i5.walmartimages.com/asr/de42310c-4cd6...   9.98          48 oz   
...                                                 ...    ...            ...   
1395  https://i5.walmartimages.com/asr/5aa587bd-9757...  13.89  not mentioned   
1396  https://i5.walmartimages.com/asr/073bdc20-ad75...  10.99         8.8 oz   
1397  https://i5.walmartimages.com/asr/242294ac-0c7d...  12.98           12 c   
1398  https://i5.walmartimages.com/asr/3a714e3e-a30e...  17.72           22 c   
1399  https://i5.walmartimages.com/asr/9f4beb5c-35c7...   2.98  not mentioned   

      weight_formatted_to_gramms  
0                         1142.5  
1                          283.5  
2                          734.3  
3                         1204.9  
4                         1360.8  
...                          ...  
1395                       272.2  
1396                       249.5  
1397                       340.2  
1398                       623.7  
1399                       623.7  

[1400 rows x 9 columns]

The data will be cleaned up by dropping rows that have NaN values in it and also the whole thumbnail column will be dropped. In addition, coffee with price of 0 will and weight of not mentioned are also be removed.

In [ ]:
df = df_raw.drop(columns='thumbnail').dropna()
df = df.loc[(df['price']>0) & (df['weight']!='not mentioned')]
df

title  \
0       folgers classic roast ground coffee, 40.3-ounce   
1     café bustelo, espresso style dark roast ground...   
2     folgers classic roast ground coffee, medium ro...   
4     great value classic roast medium ground coffee...   
5     great value classic roast medium ground coffee...   
...                                                 ...   
1390  death wish coffee, organic, fair-trade, espres...   
1392  mad priest coffee - dark night of the soul, da...   
1396  cafe yaucono espresso dark roast ground coffee...   
1397  starbucks dark roast coffee with 2x caffeine, ...   
1398  starbucks holiday blend, medium roast k-cup co...   

                     coffee_type  rating  reviews  \
0                  classic roast     3.8       93   
1            espresso,dark roast     4.7      914   
2     medium roast,classic roast     4.4      740   
4                  classic roast     4.7     1598   
5                  classic roast     4.2      263   
...                          ...     ...      ...   
1390     espresso,espresso roast     4.0        4   
1392                  dark roast     0.0        0   
1396         espresso,dark roast     0.0        0   
1397          arabica,dark roast     4.3     5036   
1398        arabica,medium roast     4.5      397   

                       seller_name  price      weight  \
0                      walmart.com  13.92  40.3-ounce   
1                      walmart.com   3.76       10 oz   
2                      walmart.com   9.97  25.9 ounce   
4                      walmart.com   9.98       48 oz   
5                      walmart.com   7.98     30.5 oz   
...                            ...    ...         ...   
1390                   walmart.com  17.97         18c   
1392                   walmart.com  12.30        10oz   
1396  coffee & filters direct, inc  10.99      8.8 oz   
1397                   walmart.com  12.98        12 c   
1398                   walmart.com  17.72        22 c   

      weight_formatted_to_gramms  
0                         1142.5  
1                          283.5  
2                          734.3  
4                         1360.8  
5                          864.7  
...                          ...  
1390                       510.3  
1392                       283.5  
1396                       249.5  
1397                       340.2  
1398                       623.7  

[967 rows x 8 columns]

In [ ]:
num_rows = df.shape[0]
num_cols = df.shape[1]

print(f'Number of rows:{num_rows}')
print(f'Number of columns:{num_cols}')

Number of rows:967
Number of columns:8


After dropping rows with NaN values in them, not mentioned as weight, and also coffee with 0 price whil also dropping the thumbnail column, the dataset has 967 rows and 8 columns

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 967 entries, 0 to 1398
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   title                       967 non-null    object 
 1   coffee_type                 967 non-null    object 
 2   rating                      967 non-null    float64
 3   reviews                     967 non-null    int64  
 4   seller_name                 967 non-null    object 
 5   price                       967 non-null    float64
 6   weight                      967 non-null    object 
 7   weight_formatted_to_gramms  967 non-null    float64
dtypes: float64(3), int64(1), object(4)
memory usage: 68.0+ KB


In [ ]:
df.head()

title  \
0    folgers classic roast ground coffee, 40.3-ounce   
1  café bustelo, espresso style dark roast ground...   
2  folgers classic roast ground coffee, medium ro...   
4  great value classic roast medium ground coffee...   
5  great value classic roast medium ground coffee...   

                  coffee_type  rating  reviews  seller_name  price  \
0               classic roast     3.8       93  walmart.com  13.92   
1         espresso,dark roast     4.7      914  walmart.com   3.76   
2  medium roast,classic roast     4.4      740  walmart.com   9.97   
4               classic roast     4.7     1598  walmart.com   9.98   
5               classic roast     4.2      263  walmart.com   7.98   

       weight  weight_formatted_to_gramms  
0  40.3-ounce                      1142.5  
1       10 oz                       283.5  
2  25.9 ounce                       734.3  
4       48 oz                      1360.8  
5     30.5 oz                       864.7

In [ ]:
df.tail()

title  \
1390  death wish coffee, organic, fair-trade, espres...   
1392  mad priest coffee - dark night of the soul, da...   
1396  cafe yaucono espresso dark roast ground coffee...   
1397  starbucks dark roast coffee with 2x caffeine, ...   
1398  starbucks holiday blend, medium roast k-cup co...   

                  coffee_type  rating  reviews                   seller_name  \
1390  espresso,espresso roast     4.0        4                   walmart.com   
1392               dark roast     0.0        0                   walmart.com   
1396      espresso,dark roast     0.0        0  coffee & filters direct, inc   
1397       arabica,dark roast     4.3     5036                   walmart.com   
1398     arabica,medium roast     4.5      397                   walmart.com   

      price  weight  weight_formatted_to_gramms  
1390  17.97     18c                       510.3  
1392  12.30    10oz                       283.5  
1396  10.99  8.8 oz                       249.5  
1397  12.98    12 c                       340.2  
1398  17.72    22 c                       623.7

In [ ]:
df.sample(10)

title  \
899   caramal machiatto single coffee cup, (regular)...   
271   dunkin’ original blend coffee, medium roast co...   
495   folgers classic roast ground coffee, medium ro...   
8     black rifle coffee tactisquatch, dark roast, g...   
886   community coffee signature blend medium roast ...   
841   black rifle coffee gunship single-serve pods, ...   
121   folgers 100% colombian coffee, medium roast gr...   
1181  cult coffee - organic house blend, whole bean,...   
1107  dunkin’ original blend coffee, medium roast co...   
370   community coffee cafe special medium roast gro...   

                         coffee_type  rating  reviews  \
899                          arabica     0.0        0   
271                     medium roast     4.8      495   
495       medium roast,classic roast     4.4      740   
8      black rifle coffee,dark roast     3.0        6   
886                     medium roast     4.9      851   
841   light roast,black rifle coffee     0.0        0   
121           colombian,medium roast     3.5        8   
1181                    medium roast     0.0        0   
1107                    medium roast     4.8      495   
370                     medium roast     4.9      381   

                  seller_name  price      weight  weight_formatted_to_gramms  
899   christopher bean coffee  16.99       18 cu                       510.3  
271               walmart.com   7.96       12 oz                       340.2  
495               walmart.com   9.97  25.9 ounce                       734.3  
8                 walmart.com  12.98       12 oz                       340.2  
886         shaan super store  24.18       32 oz                       907.2  
841               walmart.com  15.97        22 c                       623.7  
121               walmart.com  13.98  33.7 ounce                       955.4  
1181              walmart.com  12.17        10oz                       283.5  
1107              walmart.com   7.96       12 oz                       340.2  
370          jesus home depot  29.99       32 oz                       907.2

#### Top 10 Coffee Sellers

In [ ]:
df_top_sellers = df.groupby('seller_name', as_index=False).agg({
    'title': 'count'
}).rename(columns={
    'title':'product_count'
})
df_top_sellers = df_top_sellers.sort_values(by=['product_count'], ascending=False).head(10)
df_top_sellers

seller_name  product_count
90                 walmart.com            758
25     christopher bean coffee             18
51            jesus home depot             15
75           shaan super store              9
76           shaikhspeare inc.              8
55        life easy supply llc              8
63         myofficeinnovations              8
91  your voice your choice inc              7
34          door county coffee              7
87                    usa mart              7

In [ ]:
fig = px.bar(
    df_top_sellers,
    x='seller_name',
    y='product_count',
    color='seller_name',
    title='Top 10 Coffee Sellers',
    text='product_count',
    height=600,
    width=800
)
fig.show()

From the graph, we can clearly see that Walmart sells the most coffee on their stores and there are a small amount of products in Walmart stores that isn't being sold by Walmart

#### Top 10 most popular coffee by number of reviews

In [ ]:
df_reviews = df[['title', 'reviews']].sort_values(by=['reviews'], ascending=False).head(10)
df_reviews

title  reviews
46    folgers classic roast ground coffee, medium ro...    15148
968   folgers classic roast coffee, medium roast, 51 oz    14166
280   1850 black gold, dark roast coffee, k-cup pods...     7017
1303  starbucks medium roast coffee with 2x caffeine...     6301
345   gevalia mocha latte k-cup espresso coffee pods...     5315
355   1850 pioneer blend medium roast coffee, k-cup ...     5094
1397  starbucks dark roast coffee with 2x caffeine, ...     5036
385   1850 trailblazer k-cup pods, medium-dark roast...     4648
123   1850 black gold coffee, dark roast ground coff...     4165
1332  starbucks blonde roast coffee with 2x caffeine...     3835

In [ ]:
fig = px.bar(
    df_reviews,
    x='title',
    y='reviews',
    color='title',
    title='Top 10 Most Popular Coffee by Number of Reviews',
    text='reviews',
    height=900
)
fig.show()

We can see that Folgers Classic has two of the most reviewed coffee sold in Walmart stores

#### Coffee bag weight distribution

In [ ]:
fig = px.box(
    df,
    x='weight_formatted_to_gramms',
    title='Coffee bag weight distribution',
    orientation='h'
)
fig.show()

In [ ]:
fig = px.histogram(
    df,
    x='weight_formatted_to_gramms',
    title='Coffee bag weight distribution',
    width=1200,
    nbins=15
)
fig.show()

We can see that 623.7 gramms is the median and most of the coffee being sold in Walmart lie arond the ~200-399.9 gramms range. There are some outliers with weight of more than 1.6 kg.

#### Coffee Price Distribution

In [ ]:
fig = px.box(
    df,
    x='price',
    title='Coffee Price Distribution',
    orientation='h'
)
fig.show()

In [ ]:
fig = px.histogram(
    df,
    x='price',
    title='Coffee Price Distribution',
    nbins=20,
    width=1200
)
fig.show()

We can see that most of the coffee being sold in Walmart are around the range of \$10-\$14.99 and the median is \$13.09. There are some outliers with the most expensive coffee being at \$77.09

#### Coffee Type Breakdown

In [ ]:
df_coffee_type = df.groupby('coffee_type', as_index=False).agg({
    'title':'count'
}).rename(columns={
    'title':'product_count'
}).sort_values('product_count', ascending=False)
df_coffee_type

coffee_type  product_count
87                                       medium roast            275
40                                         dark roast             72
7                 black silk ground coffee,dark roast             60
38                             colombian,medium roast             54
83                                        light roast             50
..                                                ...            ...
61                                  french,dark roast              1
21                    caramel,cappuccino,medium roast              1
19      caramel flavored,caramel,arabica,medium roast              1
35                               colombian,dark roast              1
0   americano,caffe americano,medium dark,medium d...              1

[97 rows x 2 columns]

In [ ]:
fig = px.treemap(
    df_coffee_type,
    values='product_count',
    path=['coffee_type'],
    title='Coffee Type Breakdown'
)
fig.show()

We can see from the pie chart, medium roast coffee is the most that is being sold, followed by dark roast.

#### Correlation Between Price and Rating (min. 100 reviews)

In [ ]:
df_price_rating = df.loc[(df['reviews']>100) & (df['price']>0)].sort_values(by=['price'], ascending=False)
df_price_rating = df_price_rating[['price', 'rating']]
df_price_rating

price  rating
1182  77.00     4.5
1167  60.00     4.4
1221  59.89     4.7
1264  58.27     4.6
1311  54.00     4.5
...     ...     ...
381    3.72     4.4
1283   3.58     4.7
194    3.28     4.7
563    3.12     4.6
1056   2.98     4.7

[528 rows x 2 columns]

In [ ]:
fig = px.scatter(
    df_price_rating,
    x='price',
    y='rating',
    title='Correlation between Price and Rating (min. 100 reviews)',
) 
fig.show()

From the scatter plot, we can see that a lot of the coffee products being sold are around the ~\$10 range. We can also see that expensive doesn't mean good or high rated since there are coffee with higher rating but cheaper price than the coffee with the most expensive price of \$77

#### What are the most affordable coffee you can get in Walmart? (gramms per dollar value)

In [ ]:
df_affordable = df
df_affordable['gramms_per_dollar'] = df_affordable['weight_formatted_to_gramms'] / df_affordable['price']
df_affordable = df_affordable.sort_values(by=['gramms_per_dollar'], ascending=False).head(10)
df_affordable

title  \
941   international delight vanilla iced coffee, 64 oz.   
936     international delight mocha iced coffee, 64 oz.   
1027  international delight light mocha iced coffee,...   
1136  international delight reeses peanut butter cup...   
925   international delight caramel macchiato iced c...   
318   starbucks lightly sweetened premium iced coffe...   
1028  stok cold brew coffee, extra bold unsweetened,...   
945        stok cold brew coffee, not too sweet, 48 oz.   
843    stok cold brew coffee, black unsweetened, 48 oz.   
197   starbucks unweetened premium iced coffee drink...   

                        coffee_type  rating  reviews  seller_name  price  \
941                     iced coffee     4.2      586  walmart.com   4.28   
936               mocha,iced coffee     4.2      997  walmart.com   4.28   
1027              mocha,iced coffee     4.1      541  walmart.com   4.28   
1136                    iced coffee     2.5      112  walmart.com   4.28   
925   caramel,macchiato,iced coffee     4.3      635  walmart.com   4.28   
318                     iced coffee     4.8       16  walmart.com   4.57   
1028               cold brew coffee     4.8       87  walmart.com   5.28   
945                cold brew coffee     4.7      227  walmart.com   5.28   
843                cold brew coffee     4.7      367  walmart.com   5.28   
197                     iced coffee     4.2       86  walmart.com   5.28   

     weight  weight_formatted_to_gramms  gramms_per_dollar  
941   64 oz                      1814.4         423.925234  
936   64 oz                      1814.4         423.925234  
1027  64 oz                      1814.4         423.925234  
1136  64 oz                      1814.4         423.925234  
925   64 oz                      1814.4         423.925234  
318   48 oz                      1360.8         297.768053  
1028  48 oz                      1360.8         257.727273  
945   48 oz                      1360.8         257.727273  
843   48 oz                      1360.8         257.727273  
197   48 oz                      1360.8         257.727273

In [ ]:
fig = px.bar(
    df_affordable,
    x='title',
    y='gramms_per_dollar',
    title='Top 10 Most Affordable Coffee Sold in Walmart (gramms per dollar value)',
    height=800,
    color='title'
)
fig.show()

From the bar graph, we can see that International Delight produces some of the cheapest and affordable coffee sold in Walmart

#### Average Coffee Price of the Top 5 Most Popular Coffee Types

In [ ]:
df_average_price = df.groupby('coffee_type', as_index=False).agg({
    'price':'mean', 
    'title':'count'
}).rename(columns={
    'title':'product_count',
    'price':'average_price'
})
df_average_price = df_average_price.sort_values(by=['product_count'], ascending=False).head(5)
df_average_price

coffee_type  average_price  product_count
87                         medium roast      16.192800            275
40                           dark roast      15.416944             72
7   black silk ground coffee,dark roast      12.377500             60
38               colombian,medium roast      14.554444             54
83                          light roast      19.313400             50

In [ ]:
fig = px.bar(
    df_average_price,
    x='coffee_type',
    y='average_price',
    title='Average Coffee Price of Top 5 Most Popular Coffee Type',
    color='coffee_type',
    height=600,
    width=1000,
)
fig.show()

We can see that on average, light roasts are more expensive and black silk ground coffee, dark roast is less expensive

### Conclusion
- Walmart is the most popular coffee seller
- Folgers Classic produces two of the most popular coffee based on number of reviews
- The median weight is 623.7 gramms and most of the coffee being sold in Walmart lie arond the 200-399.9 gramms range. 
- The heaviest coffee sold at Walmart is 2835 gramms
- The median price is \$13.09 and most of the coffee sold in Walmart are around the \$10-14.99 price range
- The most expensive coffee sold in Walmart costs \$77.09 
- Medium Roast is the most popular roast followed by Dark Roast
- More expensive doesn't always mean better rating
- International Delight produces a lot of affordable coffee
- On average, light roasts are more expensive than medium and dark roasts. Black silk ground coffee is on average cheaper than medium roast and regular dark roasts